# Домашнее задание: Базовые типы данных



## Задание 1: Зеркальное отражение цифр

Пользователь вводит пятизначное число. Программа должна зеркально отразить три центральные цифры. Первая и последняя остаются на местах.

<br>

**Примеры:**
* `23456 -> 25436`
* `30789 -> 38709`

In [3]:
# ВАШ КОД ЗДЕСЬ

input_number = input('input 5digit number')
print(f'input number is {input_number}')
# no sense to validate digits and length, we process it as a string

first_char = input_number[0]
last_char = input_number[-1]
middle_chars = input_number[1:-1]

print(f'first: {first_char}, middle: {middle_chars}, last: {last_char}')
middle_chars = middle_chars[::-1]
changed_input = first_char + middle_chars + last_char
print(f'changed sequence is: {changed_input}')


input number is 1234ab
first: 1, middle: 234a, last: b
 changed sequence is: 1a432b


## Задание 2: Подсчет выходных до отпуска

Пользователь вводит сколько дней осталось до ближайшего отпуска. Программа должна вывести количество выходных дней до отпуска, если учесть, что выходные это суббота и воскресенье, сегодня понедельник и праздники мы не учитываем.  

<br>

**Примеры:**
* `4  -> 0`
* `6  -> 1`
* `14 -> 4`

In [30]:
# ВАШ КОД ЗДЕСЬ

def int_type_check(input_number):
    try:
        float(input_number)
        return True
    except (ValueError, TypeError):
        return False

input_number = input('enter a number of days to vacation')

while int_type_check(input_number) == False:
    input_number = input(f'it wasnt a number. Please input integer value')


days_to_vacation = int(input_number)
print(f'input days to vacation: {days_to_vacation}')

full_weeks_to_vacation = days_to_vacation // 7
days_left = days_to_vacation % 7
print(f'full weeks {full_weeks_to_vacation} and {days_left} day(s) \n---')

working_days = full_weeks_to_vacation * 5

rest_days = (full_weeks_to_vacation * 2 )
if days_left == 6:
    rest_days = rest_days + 1
    working_days = working_days + 5
    print ('if branch')
else:
    working_days = working_days + days_left

print(f'days for rest: {rest_days} \nworking days:  {working_days}')



input days to vacation: 22
full weeks 3 and 1 day(s) 
---
days for rest: 6 
working days:  16


## Задание 3: Плитка шоколада

Пользователь вводит длину и ширину плитки шоколада, а также размер куска, который хочет отломить. Программа должна вычислить - можно ли совершить подобный разлом или нет, если учесть, что ломать можно только по прямой.

<br>

**Примеры:**
* `3, 4, 6  -> True`
* `5, 7, 8  -> False`
* `4, 5, 12 -> True`

In [43]:
# ВАШ КОД ЗДЕСЬ



def int_type_check(input_number):
    try:
        int(input_number)
        return True
    except (ValueError, TypeError):
        return False

# zero and negative input validation is skipped

def check_choco_cut(choco_matrix):
    if (choco_matrix[2] % choco_matrix[0] ==0) or (choco_matrix[2] % choco_matrix[1] == 0):
        print(choco_matrix[2], choco_matrix[0], '|', choco_matrix[1] )
        print(f'{choco_matrix[2] % choco_matrix[0]}, {choco_matrix[2] % choco_matrix[1]}')
        return True
    else:
        return False


# variables:
input_instructions = ['enter choco width as integer', 'enter choco length as integer', 'enter desired tiles amount as integer']
choco_matrix = []

# prompt user to enter choco dimensions and store them as a list
for x in range(0,3):
    input_number = input(f'{input_instructions[x]}')

    while int_type_check(input_number) == False:
        input_number = input(f'it wasnt a number. Please input integer value')
    choco_matrix.append(int(input_number))

print(f'choco dimensions: {choco_matrix}')
print(f'User wanna cut {choco_matrix[2]} tiles')

# Check whether its possible to cut the desired tiles rectangle
print(f' is it possible? {check_choco_cut(choco_matrix)}')



choco dimensions: [4, 5, 8]
User wanna cut 8 tiles
8 4 | 5
0, 3
 is it possible? True


## Задание 4: Римские числа

Пользователь вводит целое положительное число, программа должна вернуть строку в виде римского числа.

<br>

**Примеры:**
* `3   -> III`
* `15  -> XV`
* `234 -> CCXXXIV`

In [62]:
# ВАШ КОД ЗДЕСЬ


#constants
roman_ones =    ['', 'I', 'II', 'III', 'IV', 'V', 'VI', 'VII', 'VIII', 'IX', 'X']
roman_tens =    ['', 'X', 'XX', 'XXX', 'XL', 'L', 'LX', 'LXX', 'LXXX', 'XC']
roman_hundreds= ['', 'C', 'CC', 'CCC', 'CD', 'D', 'DC', 'DCC', 'DCCC', 'CM']
roman_thousands=['', 'M', 'MM', 'MMM']

def get_char_position(input_number, position):
    if 0 <= position < len(input_number):
        return input_number[position]
    else:
        return None

def make_int_if_not_none(value):
    if value is None or value == '':
        return 0
    return int(value)

def roman_view_transform(thousands_part, hundreds_part, tens_part, ones_part, roman_view=None):
    if thousands_part != None:
        roman_view = str(roman_thousands[thousands_part])
    if hundreds_part != None:
        roman_view = str(roman_view + (roman_hundreds[hundreds_part]))
    if tens_part != None:
        roman_view = roman_view + (roman_tens[tens_part])
    if (ones_part != None) or (ones_part > 0):
        roman_view = roman_view + (roman_ones[ones_part])
    return roman_view


input_number = input('input a positive number as integer 1 - 3000')
print(f'input number: {input_number}')
input_number = input_number[::-1]
print(f'reversed input number: {input_number}')


ones_part = make_int_if_not_none(get_char_position(input_number, 0))
tens_part = make_int_if_not_none(get_char_position(input_number, 1))
hundreds_part = make_int_if_not_none(get_char_position(input_number, 2))
thousands_part = make_int_if_not_none(get_char_position(input_number, 3))


print(f'roman view: {roman_view_transform(thousands_part, hundreds_part, tens_part, ones_part, roman_view='')}')



input number: 299
reversed input number: 992
roman view: CCXCIX


## Задание 5: Проверка на вещественное число

Пользователь вводит данные, проверить - являются ли они положительным вещественным числом. Не использовать встроенные функции для проверки, только методы данных и конструкцию IF. (Дополнительное задание, по желанию - проверка на отрицательные вещественные числа)

<br>

**Примеры:**
* `5.6  -> True`
* `.78  -> True`
* `.67. -> False`
* `5    -> True`

In [156]:
# ВАШ КОД ЗДЕСЬ

input_number = input('enter your data to check whether its a "real number"')

# remove spaces
spaces = 0
for char in input_number:
    if char == ' ':
        spaces +=1
        input_number = input_number.replace(' ', '')
if spaces > 0:
    print(f'removed spaces, now input number is: {input_number}')


def check_dot_amount(input_number):
    dots = 0
    for char in input_number:
        if char == '.':
            dots +=1
    if dots > 1:
        print(f'dots: {dots} ; only 1 dot allowed for real numbers')
    return dots

def check_first_minus_char(input_number):
    if input_number[0] == '-':
        return 'negative'
    else:
        return 'positive'

def check_minuses(input_number):
    minuses = 0
    for x in range(1,len(input_number)):
        if input_number[x] == '-':
            minuses +=1
            if minuses > 0:
                print(f'only 1 first "-" allowed for real numbers')
    return minuses


def check_wrong_symbols(input_number):
    wrong_symbols = 0
    for char in input_number:
        if (char != '.') and (char != '-') and not ('0' <= char <= '9'):
            wrong_symbols +=1
    if wrong_symbols >0:
        print(f'wrong symbols: {wrong_symbols} ; use only digits ,"-" and "."  \nNO letters, carriage return or other special symbols')
    return wrong_symbols


if (check_wrong_symbols(input_number) == 0) and (check_dot_amount(input_number) < 2) and (check_minuses(input_number) < 1):
    if check_first_minus_char(input_number) == 'negative':
        print(f'"{input_number}" is a negative real number')
    else:
        print(f'"{input_number}" is a positive real number')
else:
    print(f'"{input_number}" is NOT a real number')



removed spaces, now input number is: -123.4
"-123.4" is a negative real number
